File: golang/ds/maxHeap.go
package ds

import "fmt"

type MaxHeap struct {
	container []int
}

func (h *MaxHeap) Push(val int) {

	h.container = append(h.container, val)
	h.heapifyUp(len(h.container) - 1)
}

func (h *MaxHeap) Pop() {
	if len(h.container) == 0 {
		fmt.Println("Empty...")
		return
	}

	h.swap(0, len(h.container)-1)
	h.container = h.container[:len(h.container)-1]
	h.heapifyDown(0)
}

func (h *MaxHeap) Peek() int {
	if len(h.container) == 0 {
		return -1
	}
	return h.container[0]
}

func (h *MaxHeap) getParentIdx(idx int) int {

	return (idx - 1) / 2
}

func (h *MaxHeap) getLeftChildIdx(idx int) int {

	return 2*idx + 1
}

func (h *MaxHeap) getRightChildIdx(idx int) int {

	return 2*idx + 2
}

func (h *MaxHeap) withinBounds(idx int) bool {
	return idx >= 0 && idx < len(h.container)
}
func (h *MaxHeap) heapifyUp(idx int) {

	if idx == 0 {
		return
	}

	parentIdx := h.getParentIdx(idx)
	if !h.withinBounds(parentIdx) {
		return
	}
	parentVal := h.container[parentIdx]
	currVal := h.container[idx]
	if currVal > parentVal {
		h.swap(parentIdx, idx)
		h.heapifyUp(parentIdx)
	}

}

func (h *MaxHeap) swap(idx1 int, idx2 int) {
	h.container[idx1], h.container[idx2] = h.container[idx2], h.container[idx1]
}

func (h *MaxHeap) PrintContainer() {
	for _, val := range h.container {
		fmt.Printf("\nValue is %d", val)
	}
	fmt.Println()
}

func (h *MaxHeap) heapifyDown(idx int) {

	if !h.withinBounds(idx) {
		return
	}
	lIdx := h.getLeftChildIdx(idx)
	rIdx := h.getRightChildIdx(idx)

	if h.withinBounds(lIdx) && h.withinBounds(rIdx) {

		if h.container[lIdx] > h.container[rIdx] && h.container[lIdx] > h.container[idx] {
			h.swap(lIdx, idx)
			h.heapifyDown(lIdx)
		} else if h.container[rIdx] > h.container[lIdx] && h.container[rIdx] > h.container[idx] {
			h.swap(rIdx, idx)
			h.heapifyDown(rIdx)
		}
	} else if h.withinBounds(lIdx) {

		if h.container[lIdx] > h.container[idx] {
			h.swap(lIdx, idx)
			h.heapifyDown(lIdx)
		}

	} else if h.withinBounds(rIdx) {
		if h.container[rIdx] > h.container[idx] {
			h.swap(rIdx, idx)
			h.heapifyDown(rIdx)
		}
	}

}

File: typescript/maxHeap.ts
export default class MaxHeap {

    container: Array<number>

    constructor(){
        this.container = []
    }

    private getParent(idx: number){
        return Math.max(0,Math.floor((idx-1)/2))
    }

    private swap(idx1:number,idx2:number):void {
        const tmp = this.container[idx1]
        this.container[idx1] = this.container[idx2]
        this.container[idx2] = tmp
    }

    private heapifyUp(idx: number): void{
        if(this.container.length === 0 || this.container.length === 1){
            return
        }
        const pIdx = this.getParent(idx) 
        const p = this.container[pIdx] 
        const c =this.container[idx]
        if(c > p) {
            this.swap(idx,pIdx)
            this.heapifyUp(pIdx)
        }

    }
    private heapifyDown(idx:number): void{
        if(idx >= this.container.length - 1){
            return
        }
        const lcIdx = 2*idx + 1
        const rcIdx = 2*idx + 2
        let mcIdx = 0
        if(this.container[lcIdx] && this.container[rcIdx]){
            mcIdx = this.container[lcIdx] > this.container[rcIdx] ? lcIdx : rcIdx
        }
        else if(this.container[lcIdx] || this.container[rcIdx]){
            mcIdx = this.container[lcIdx] ? lcIdx : rcIdx
        }
        else{
            return
        }

        if(this.container[mcIdx] > this.container[idx]) {
            this.swap(mcIdx,idx)
        }
        this.heapifyDown(mcIdx)


    }

    push(value: number): void{
        this.container.push(value)
        this.heapifyUp(this.container.length-1)
    }

    pop(): Number | undefined{
        if(this.container.length === 0){
            return undefined
        }
        const val = this.container[0]
        this.swap(this.container.length-1,0)
        this.container.pop()
        this.heapifyDown(0)
        return val
    }

    peek(): Number | undefined {
        return this.container[0]
    }


}
